Replace alert usage with tgui_alert

celotajstg opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Pretty much every alert() call is replaced with tgui_alert, except one I replaced with tgalert as a fallback. If tgui_alert exists, why not use it?

## Why It's Good For The Game

More consistent user interface, as well as just the fact it looks better.

Before:
![image](https://user-images.githubusercontent.com/81999976/114690502-6e990080-9d1f-11eb-9eaa-f1d4600a11d2.png)
After:
![image](https://user-images.githubusercontent.com/81999976/115035588-6682c180-9ed5-11eb-919d-f028c207a490.png)

## Changelog
:cl: Celotajs, stylemistake
refactor: Almost all alert calls have been replaced with its tgui alternative, tgui_alert
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

celotajstg committed:
Replace alert usage with tgui_alert

optimumtact commented:
What about those who cannot run tgui successfully hmm?
![image](https://user-images.githubusercontent.com/2544323/114690880-3d462200-9d6b-11eb-9508-a2c284385c12.png)

celotajstg commented:
The game is already unplayable if you cannot run tgui anyway

celotajstg commented:
Also the tools label should be removed, I removed the change that affects a tool

stylemistake commented:
Can you wrap Alerts in a Section while you're at it for higher contrast, thanks.

stylemistake commented on code/modules/tgui_panel/external.dm:
```diff
@@ -16,8 +16,8 @@
 
-	// Failed to fix
-	action = alert(src, "Did that work?", "", "Yes", "No, switch to old ui")
+	// Failed to fix, using tgalert as fallback
+	action = tgalert(src, "Did that work?", "", "Yes", "No, switch to old ui")
```
Keep such things as "fix verbs" using old alerts because even when everything breaks, these flows should still work.

celotajstg replied:
Just want to clarify - in the PR where tgalert has been replaced with tgui_alert, tgalert was kept as a fallback for the report issue button, so why should tgalert be replaced with alert here?

stylemistake replied:
Good question, maybe it is robust indeed.

SplinterGP replied:
TGalert aint still html? It probably works as good as alert()

celotajstg committed:
Improve contrast in the TGUI alert modal

celotajstg commented:
Added the section wrap:
![image](https://user-images.githubusercontent.com/81999976/114694352-1d8b0b80-9d23-11eb-95db-47acdae37d30.png)

Fikou commented:
are these timed or what are those bars at the top

celotajstg commented:
Yes, 60 seconds by default

celotajstg commented:
Also if everything will be fine I'd like to request this PR to be testmerged first

celotajstg committed:
Update alert modal styling

celotajstg committed:
More replacements, some with tgui_input_list

MrStonedOne commented:
Do not make alerts timed that were not timed before. Especially in admin tools like vv_do_topic

MrStonedOne commented:
Some users have an issue where browser windows randomly take 1 minute or two to open.

I'm users.

tgui_alert is a replacement for tg_alert, it is not a replacement for alert, at least not while the timer defaults to 60 seconds.

MrStonedOne commented on code/modules/client/client_procs.dm:
```diff
@@ -328,7 +328,7 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 	if(alert_mob_dupe_login)
 		spawn()
-			alert(mob, "You have logged in already with another key this round, please log out of this one NOW or risk being banned!")
+			tgui_alert(mob, "You have logged in already with another key this round, please log out of this one NOW or risk being banned!",,list("Ok"))
```
Another example of an alert that must not be timed. We **ban** people based on the assumption they have seen this alert.

celotajstg commented:
Got it, I'll look into making those affected in this ticket non-timed

celotajstg commented:
I have adjusted current behavior to be the same as alert() behavior, meaning that by default a tgui_alert is not closable, without a timeout, and with a single "Ok" button unless specified otherwise. This will make everything work as it previously did, however just with a better appearance.

Here is how the "Observe" button will look - same behavior as the old alert.
![image](https://user-images.githubusercontent.com/81999976/114948410-98edd980-9e57-11eb-9ce4-94604b2844a1.png)

celotajstg committed:
Adjust default tgui_alert behavior

celotajstg committed:
Resolve linting issues

willox commented:
Do these windows take keyboard focus like the current alert does? I think quite a lot of people probably use the windows with their arrow keys/enter.

celotajstg commented:
Yep, they do, just tesetd on the observer alert.

celotajstg commented:
Just to clarify - you can cycle through buttons with tab/shift tab and select them with space or enter.

Timberpoes commented:
I have issues where tgui windows will, from time to time, whitescreen and self close.

Tgui is reliable enough for general use and I'll only have whitescreen interfaces that timeout and close a couple of times a shift, but tgui popups that were not opened by an input from the user give me usability concerns in general.

If the user has no way to reopen the alert from a user input, it probably should remain non-tgui until we figure out a fix to whitescreen tgui interfaces.

This is not a maintainer blocking comment, more my experience as an end user where I've had tgui prompts pop up as whitescreen and then self close as a result.

celotajstg commented:
I think it would be worth trying to testmerge it and collect some feedback, which is pretty much why I am requesting not to instantly just merge.

celotajstg committed:
Rename noClose to canClose, adjust behavior

celotajstg commented:
Todo - minor styling improvements

stylemistake commented on tgui/packages/tgui/layouts/Window.js:
```diff
@@ -24,9 +24,15 @@ const DEFAULT_SIZE = [400, 600];
     if (suspended) {
       return;
     }
+    if (!canClose) {
+      Byond.winset(window.__windowId__, {
+        'can-close': false,
+      });
+    }
```
This is bad, because this window will never restore to `can-close: true` once it's set to `false`.

This would be much better, even though it is slightly more expensive to always call it, but it is possible to avoid the call by storing the currently set value as a variable, and only call it when it changes.

```jsx
Byond.winset(window.__windowId__, {
  'can-close': Boolean(canClose),
});
```

celotajstg committed:
Improve accessibility for alerts

Co-authored-by: actioninja <actioninja@criticalaction.net>

celotajstg commented:
Accessbility has been improved - it is now possible to change buttons using arrow keys in addition to tab / shift tab, and the first button is focused by default. The focused button is also going to be slightly more contrasting than previously.
![image](https://user-images.githubusercontent.com/81999976/115035588-6682c180-9ed5-11eb-919d-f028c207a490.png)

Alt+F4 is also now unusable for closing the alert dialog.

KathyRyals commented:
I'd just like to say I am impressed by the effort and consistent improvements the OP has made to this PR. Regardless of the eventual merge/close, I applaud them for taking feedback so well and spending so much time on improving our UI and UX. Bravo.

stylemistake committed:
Update Window.js

stylemistake commented:
Highly commendable work, @celotajstg!

Mothblocks commented:
I have some minor UX problems with this, but it's been mostly stable.

Two things:
1. I like to click Observe, wait for the round to setup, then press Enter to complete the process. It looks like the observe window doesn't properly focus to where this can happen.
2. I was using a holopad the other day, and used my keyboard to press the "R" key until I got to a certain spot. Every time I did, it correctly went to the next input, but *also* toggled throw mode.

Fikou commented:
theres some places where the tgui alerts feel really weird, like in admin sound playing procs where it switches between the tgui and plain white stuff for picking from a list/setting an url/whatever
a few admins had a problem where they were pressing a button when for example a ghost role popped up and they got put in some random animal
add reagent in vv dropdown seems to be broken

SplinterGP commented:
On TGMC we have basically almost all alerts except admin specific stuff that we want max reliability as tgui_alert and tgui_input_list and they work fine

MrStonedOne commented:
> On TGMC we have basically almost all alerts except admin specific stuff that we want max reliability as tgui_alert and tgui_input_list and they work fine, even without some qol stuff from this and other stuff i dindt pull yet

Things like that can self select against the random users who get bugs.

The issue here is we have a number of admins and players with a bug where html windows will randomly whitescreen for 15 to 120 seconds when opening, normally this only happens on the first window opened in a given DS process (outside of the lobby prompt as that seems mostly immune). this gets worse with tgui because it will often close the window after a delay if it doesn't load.

SplinterGP commented:
Thats odd, on TGMC similar happens to me when i boot dev server(some completely white window appears) and sometimes some UI's from some stuff ceases to open, like, i open a vendor UI sometimes it gets closed by getting away or i close and i cant re-open the UI of that vendor and its fairly common to happen, althought with alert seemed fine since i think the problem is with a particular UI of a datum/atom

MrStonedOne commented:
> althought with alert seemed fine since i think the problem is with a particular UI of a datum/atom

Na, the bug is in byond. it was introduced in 509 on like the second to last release of the beta and its something to do with byond's initialization of the ie window being delayed by \*something\*. its random what computers it happens to or when it happens, but seems to be more likely to happen if the computer locale/codepage was ever changed (such as changing it to Japanese to run older japanese games) even if its later changed back.

celotajstg commented:
I think the initialization might be improved in 514, so I believe it will be worth doing another test once we get there.

SplinterGP commented:
Ah yes byond and windows, what a couple, well good to know.

Mothblocks commented:
When you type `:)` in OOC it blue screens, it's the prompt you get when it thinks you're trying to send a radio.

celotajstg commented:
Rebased, will proceed with checking for bugs and issues soon

celotajstg committed:
Fix add reagents

celotajstg committed:
Fix meant for OOC alert

celotajstg committed:
Stop alert windows from fighting for focus

celotajstg commented:
> add reagent in vv dropdown seems to be broken
> When you type :) in OOC it blue screens, it's the prompt you get when it thinks you're trying to send a radio.

Fixed. The errors in OOC and add reagents were due to an oversight I made by not converting some choices to a list parameter, and after reading every single line change I made I've come to conclusion that, luckily, those were the only oversights.

> a few admins had a problem where they were pressing a button when for example a ghost role popped up and they got put in some random animal

Fixed. My current way of focusing the alert window was too intrusive, meaning that if you happened to switch to the chat or something, it would force the focus back on itself.

Now if you want to use keyboard navigation in the alert again, you need to either click on the title bar and hover over the window contents, or click the window contents. Windows now don't fight each other for focus.

> I was using a holopad the other day, and used my keyboard to press the "R" key until I got to a certain spot. Every time I did, it correctly went to the next input, but also toggled throw mode.

I believe this might be best to handle this in another PR regarding how TGUI treats button capturing in this case.

> theres some places where the tgui alerts feel really weird, like in admin sound playing procs where it switches between the tgui and plain white stuff for picking from a list/setting an url/whatever

This is as it should be for now, I'm keeping this PR to mostly alerts. That's a job for another PR.



Overall, I think we can merge most likely.

celotajstg commented:
Rebased to fix yet another conflict.

Mothblocks commented:
just do a normal merge commit

Cyberboss commented:
The PR is getting squashed anyway

optimumtact commented:
rebasing is superior

celotajstg committed:
Sentience baloon

celotajstg commented:
ended up having to cherry-pick due to my html cleanup pr leading to git shenanigans, but overall nothing new aside from me accidentally committing a merge conflict and then fixing it
